File: 0039-组合总和/Solution.java
import java.util.*;


class Solution {
    private List<List<Integer>> res = new ArrayList<>();
    private int[] candis;
    private int length;

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        candis = candidates;
        length = candidates.length;

        if(length == 0){
            return res;
        }
        
        Arrays.sort(this.candis);
        Stack<Integer> path = new Stack<>();

        helper(0, target, path);
        return res;
    }


    private void helper(int start, int target, Stack<Integer> path){
        if(target == 0){
            this.res.add(new ArrayList<>(path));
            return;
        }

        for(int i=start; i<this.length; i++){
            int result = target - this.candis[i];
            if(result < 0){
                continue;
            }
            path.push(this.candis[i]);
            helper(i, result, path);
            path.pop();
        }
    }
}

File: 0039-组合总和/Solution.py
class Solution:
    def __init__(self):
        self.cache = {0: [[]]}

    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ret = []
        candidates.sort()
        self.dfs(candidates, 0, [], ret, target)
        return ret

    def dfs(self, candidates, index, path, ret, target):
        if target < 0:
            return
        if target == 0:
            ret.append(path)
        for i in range(index, len(candidates)):
            if target < candidates[i]:
                break
            self.dfs(candidates, i, path+[candidates[i]], ret, target-candidates[i])

    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        ret = self.trace(candidates, target)
        result = []
        for path in ret:
            path.sort()
            if path not in result:
                result.append(path)
        return result

    def trace(self, candidates, target):
        ret = []
        if target in self.cache:
            return [v[:] for v in self.cache[target]]
        for num in candidates:
            if target < num:
                break
            tmp = self.combinationSum(candidates, target-num)
            for i in range(len(tmp)):
                tmp[i].append(num)
            ret.extend(tmp)
        self.cache[target] = [v[:] for v in ret]
        return ret

